Source/Core/DolphinQt2/Config/ControllersWindow.cpp

```diff
@@ -0,0 +1,349 @@
+
+  m_main_layout->addWidget(m_advanced_box);
+
+  setLayout(m_main_layout);
+}
+
+void ControllersWindow::ConnectWidgets()
+{
+  connect(m_wiimote_passthrough, SIGNAL(toggled(bool)), this, SLOT(OnWiimoteModeChanged(bool)));
+
+  connect(m_advanced_bg_input, SIGNAL(toggled(bool)), this, SLOT(SaveSettings()));
+  connect(m_wiimote_continous_scanning, SIGNAL(toggled(bool)), SLOT(SaveSettings()));
+  connect(m_wiimote_real_balance_board, SIGNAL(toggled(bool)), SLOT(SaveSettings()));
+  connect(m_wiimote_speaker_data, SIGNAL(toggled(bool)), SLOT(SaveSettings()));
+  connect(m_wiimote_sync, SIGNAL(clicked()), SLOT(OnBluetoothPassthroughSyncPressed()));
+  connect(m_wiimote_reset, SIGNAL(clicked()), SLOT(OnBluetoothPassthroughResetPressed()));
+
+  for (int i = 0; i < 4; i++)
+  {
+    connect(m_wiimote_boxes[i], SIGNAL(currentIndexChanged(int)), SLOT(SaveSettings()));
+    connect(m_wiimote_boxes[i], SIGNAL(currentIndexChanged(int)), SLOT(OnWiimoteTypeChanged(int)));
+    connect(m_wiimote_buttons[i], SIGNAL(clicked()), SLOT(UnimplementedButton()));
+    connect(m_gc_controller_boxes[i], SIGNAL(currentIndexChanged(int)), SLOT(SaveSettings()));
+    connect(m_gc_controller_boxes[i], SIGNAL(currentIndexChanged(int)), SLOT(OnGCTypeChanged(int)));
+    connect(m_gc_buttons[i], SIGNAL(clicked()), SLOT(UnimplementedButton()));
+  }
+}
+
+void ControllersWindow::OnWiimoteModeChanged(bool passthrough)
+{
+  m_wiimote_sync->setEnabled(passthrough);
+  m_wiimote_reset->setEnabled(passthrough);
+
+  for (int i = 0; i < 4; i++)
+  {
+    m_wiimote_boxes[i]->setEnabled(!passthrough);
+    m_wiimote_buttons[i]->setEnabled(!passthrough);
+  }
+
+  m_wiimote_real_balance_board->setEnabled(!passthrough);
+  m_wiimote_speaker_data->setEnabled(!passthrough);
+  m_wiimote_continous_scanning->setEnabled(!passthrough);
+
+  SaveSettings();
+}
+
+void ControllersWindow::OnWiimoteTypeChanged(int type)
+{
+  auto* box = static_cast<QComboBox*>(QObject::sender());
+
+  for (int i = 0; i < 4; i++)
+  {
+    if (m_gc_controller_boxes[i] == box)
+    {
+      m_wiimote_buttons[i]->setEnabled(box->currentIndex() != 0);
+      return;
+    }
+  }
+}
+
+void ControllersWindow::OnGCTypeChanged(int type)
+{
+  auto* box = static_cast<QComboBox*>(QObject::sender());
+
+  for (int i = 0; i < 4; i++)
+  {
+    if (m_gc_controller_boxes[i] == box)
+    {
+      m_gc_buttons[i]->setEnabled(box->currentIndex() != 0);
+      return;
+    }
+  }
+}
+
+void ControllersWindow::OnBluetoothPassthroughResetPressed()
+{
+  const auto ios = IOS::HLE::GetIOS();
+
+  if (!ios)
+  {
+    QMessageBox error(this);
+    error.setIcon(QMessageBox::Warning);
+    error.setText(tr("Saved Wii Remote pairings can only be reset when a Wii game is running."));
+    error.exec();
+    return;
+  }
+
+  auto device = ios->GetDeviceByName("/dev/usb/oh1/57e/305");
+  if (device != nullptr)
+  {
+    static_cast<IOS::HLE::Device::BluetoothBase*>(device.get())->TriggerSyncButtonHeldEvent();
```

Comment:
Since device is a shared_ptr, it can utilize [std::static_pointer_cast](http://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast) which makes the `.get()` unnecessary.